programs/server/config.xml

```diff
@@ -272,16 +271,67 @@
+                            Accepted values are: 'base', 'one_level', 'children', 'subtree' (the default).
+                    search_filter - template used to construct the search filter for the LDAP search.
+                            The resulting filter will be constructed by replacing all '{username}', '{bind_dn}', and '{base_dn}'
+                             substrings of the template with the actual user name, bind DN, and base DN during each LDAP search.
+                            Note, that the special characters must be escaped properly in XML.
+                    fail_if_all_rules_mismatch - flag to trigger failure if none of the rules were able to match and transform any
+                             of the resulting strings returned by the LDAP search. By default, set to 'true'.
+                    rule - section with matching and mapping info.
+                            Each string will be matched to the regex and, if there is a full match, replaced using format string to
+                             get the name of the local role that needs to be assigned to the user.
+                            There can be multiple 'rule' sections defined inside the same 'role_mapping' section. Each of those rules
+                             will be applied in the order they are listed in 'role_mapping' section. If a rule does not match a string
+                             the next rule will be applied. If none of the rules matched a string and 'fail_if_all_rules_mismatch' is
+                             set to 'false', that particular string will be ignored. If a rule matched a string and 'continue_on_match'
+                             is set to 'false', the subsequent rules will not be applied to the current string.
+                        match - regular expression, in ECMAScript format, used to match each entire string retured by LDAP serach. If
+                             matched successfully, a replacement will be performed and the resulting string will be treated as local
+                             role name and assigned to the user. By default, set to '.+' (match any non-empty string).
+                            Note, that the special characters must be escaped properly in XML.
+                        replace - format string used as a replace expression after the match succeeded. References like '$&' (entire
+                             matched string), or '$n' (n-th subgroup) can be used. By default, set to '$&'.
+                            Note, that the special characters must be escaped properly in XML.
+                        continue_on_match - flag that indicates, whether to continue matching and mapping using the subsequent rules
+                             after this rule successfully matched and mapped the string. By default, set to 'false'.
+                            If set to 'true' and multiple rules match, multiple role names may be generated from one same input string.
```

Comment:
These rules look quite complex. Did you elaborate them by yourself or took from somewhere?

Reply:
No, this is not an adoption of something known. This is meant to be the most concise and generic way of transforming names, that covers any scenario we could imagine so far. The description may not be the best, but the essence of the rules is simple: chain of match+replace rules, where each of them produces a role name, if able to match the input string.

Reply:
Pushed a fix.

Reply:
It would better be an adoption ;)

Those regexp rules seem to be too difficult. And two extra boolean flags look you're trying to invent your own string processing programming language. And xml escaping definitely doesn't make the things easier.

Is it really necessary? I mean, just `search_filter`, `scope` and `attribute` are not enough? If not maybe one string-replacement rule can solve the problem. But if you're sure that we need a complex system with lots of replacement rules and control flow then I suppose it's better to enable something like AWK scripting.

Reply:
> It would better be an adoption ;)

Most of the systems don't have matching rules, they just assign names as-is. Clearly less flexible.

> Those regexp rules seem to be too difficult.

That's a generic regexp. Two fields, one for match and one for replace. It's hard for me to imagine anything simpler. :)
Implementing AWK would require the same number of config fields, but much more hustle in the code. And will be of same mental complexity as regexps, while less common.

> maybe one string-replacement rule can solve the problem.

Really, the same number of config fields, and much less flexible. And we may not solve the actual problems of our customers after all.

> Is it really necessary?

The thing is, there are many different environments, and we don't have all the requirements. So implementing something that will cover virtually all of the cases seems to be optimal way to spend resources.

Reply:
The fields have reasonable default values, so no need to specify all of them, to achieve typical behavior (of other systems).

Reply:
AWK is more general cause it cover almost everything which one can imagine about string processing. And it requires less fields in the config because instead of whole bunch of `<rule>`, `<match>`, `<replace>`, `<fail_if_all_rules_mismatch>`, `<continue_on_match>` you will have to specify only a path to *.awk file. I don't insist on using AWK though however it seems to be a decent alternative.

Reply:
It would be harder to integrate, and performance impact would be noticeable. But we can add it too, if the approach is proven to be more convenient.

Reply:
> harder to integrate

I agree, it would require an extra library in our `contrib` and it doesn't seem necessary right now.

Reply:
> That's a generic regexp. Two fields, one for match and one for replace. It's hard for me to imagine anything simpler. :)

For example, you cannot make string lowercased or uppercased with regexps. You cannot extract substring by indices. And you have troubles even writing those regexps with all that xml escaping.

If you like regexps, let's keep them. But I believe we should do something about those boolean flags, they're hard to explain and hard to understand. I suggest making `<apply_all_rules>` as a replacement for  `continue_on_match=true`, ..., `continue_on_match=true` and `<apply_any_rule>` as a replacement for `continue_on_match=false`, ..., `continue_on_match=false`. And `fail_if_all_rules_mismatch` doesn't seem to be very necessary at all: if no rules match then probably role will not be found and so user will get his error message anyway.

I just want this feature to be usable. That's why I think too much complexity is bad.

Reply:
I don't see how this is more simple. Same two flags, less flexibility. Regarding to-lower/to-upper, since we never heard about such use case, let's skip it until requested. Let's focus on a first iteration, then we will adjust it based on a feedback, not on our guesses.

By the way, it makes no sense for role names to be case sensitive.

Reply:
It's more about usability than programming. I feel we maybe should ask someone else. @filimonov can you say something about it?

Reply:
I like simple things. Conventions over complicated configurations. Less flexible but 100% clear and predictable.

I would be happy with something like: if user will have a group with 'clickhouse_' prefix clickhouse will assign such a role (w/o prefix) to the user. Simple to explain, you don't need to look 15 mins in configurations to guess/understand what kind of privilegues the particular user will get after applying all those regexps. 

I have no clear understanding what advantages those complicated rules give (and I'm bit scared of how i will need to explain that to somebody) 

Maybe we can go with simple setup with predefined / simple matching rule at the beginning, and start extending it when smbd will came with  such a requirement

Reply:
Changed to prefix.